### ISSUE TITLE
Timber::get_post returns a Timber\Post object even when the post ID doesn't exist

### ISSUE THREAD

**Issue by chrisgherbert** [timber/timber#1467]
### Expected behavior
`Timber::get_post()` on a non-existent post ID should return `null` (like the native WordPress `get_post()` function) or at least some kind of falsey value to allow easy conditionals.

### Actual behavior
A less-than-useful, mostly empty Timber\Post object is returned. For example:

```
>>> Timber::get_post('1312312');
=> Timber\Post {#1461
     +ImageClass: "Timber\Image",
     +PostClass: "Timber\Post",
     +TermClass: "Timber\Term",
     +object_type: "post",
     +class: "",
     +id: null,
     +ID: "1312312",
     +post_author: null,
     +post_content: null,
     +post_date: null,
     +post_excerpt: null,
     +post_parent: null,
     +post_status: null,
     +post_title: null,
     +post_type: null,
     +slug: null,
   }
```

### Steps to reproduce behavior
Call `Timber::get_post()` on a nonexistent post ID.  It does seem to check if the value is numeric -- a non-numeric string returns `null`.

### What version of WordPress, PHP and Timber are you using?
WordPress 4.8, PHP 5.6, Timber 1.3.2

### How did you install Timber? (for example, from GitHub, Composer/Packagist, WP.org?)
Installed using Composer

**Comment by twentyfortysix** [comment]
+1

**Comment by gchtr** [comment]
Totally agree with you! I already saw that quite a few times myself.

Currently, `Timber::get_post()` is marked as deprecated for v2.0 of Timber. But that would be a good reason to keep the `get_post()` function, because when `new Timber\Post()` is the only way to create a new object, we can’t return `null`.

I only see three ways to handle this:

1. Throw an exception.
2. Have a custom method on a `Timber\Post` object, that could be called to check whether it’s a valid post. For example: `$post->is_valid()`, or `$post->exists()`.
3. Prevent the post object from initializing itself and maybe return `null` for every method and property that is accessed.
4. Keep `get_post()`.

Option 1 doesn’t seem very applicable to me. I guess we always strive to keep the public-facing API simple. Exceptions would be a new concept that we don’t have in Timber yet, right?

Option 2 could also be combined with Option 3, so that at least no errors are thrown.

Are there other options? What’s your opinion, @jarednova?

**Comment by jarednova** [comment]
@chrisgherbert and @twentyfortysix are right-on the money. This was a bad oversight in the original construction that's unfortunately lived on to this day. 

Of the options laid out by @gchtr I'd vote for something like number 3. The spirit of the WP result is "null" so I think anything that carries this through (eg. `title=null`, `ID=null`, `content=null`...) would take advantage of both existing properties/methods and the WP expectation.

I could be talked into 2, but I'd prefer to start in a place that doesn't require a new public method (thus, new docs, new support reqs., new potential custom field conflicts...). None of those is major on its own, but I'd rather go there as a last (or at least, second) resort

**Comment by acobster** [comment]
@jarednova @gchtr I'm running into some strangeness trying to run the `test-timber-post.php` unit tests:

```
vagrant@vvv (2.x) /vagrant/www/timber $ phpunit tests/test-timber-post.php                                                                                                           [19/1925]
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
Use http://build.starter-theme.dev/ for testing with UI
PHPUnit 5.7.16 by Sebastian Bergmann and contributors.

Error:         No code coverage driver is available

...E.E........................................................... 65 / 72 ( 90%)
...F.EF                                                           72 / 72 (100%)

Time: 4.42 seconds, Memory: 30.00MB

There were 3 errors:

1) TestTimberPost::testGetImage
Error: Class 'TestTimberImage' not found

/vagrant/www/timber/tests/test-timber-post.php:30
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:186
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:116

2) TestTimberPost::testFalseParent
Error: Class 'TestTimberImage' not found

/vagrant/www/timber/tests/test-timber-post.php:49
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:186
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:116

3) TestTimberPost::testPathAndLinkWithPort
Error: Class 'TimberPost' not found

/vagrant/www/timber/tests/test-timber-post.php:960
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:186
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:116

--

There were 2 failures:

1) TestTimberPost::testPostWithVideo
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '<iframe width="500" height="281" src="https://www.youtube.com/embed/Jf37RalsnEs?feature=oembed" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>'
+    0 => '<iframe width="500" height="281" src="https://www.youtube.com/embed/Jf37RalsnEs?feature=oembed" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture
-in-picture" allowfullscreen></iframe>'
 )

/vagrant/www/timber/tests/test-timber-post.php:937
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:186
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:116

2) TestTimberPost::testEditUrl
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'http://example.org/wp-admin/post.php?post=1&amp;action=edit'
+'http://example.org:3000/wp-admin/post.php?post=1&amp;action=edit'

/vagrant/www/timber/tests/test-timber-post.php:989
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:186
/usr/local/src/composer/vendor/phpunit/phpunit/src/TextUI/Command.php:116

ERRORS!
Tests: 72, Assertions: 113, Errors: 3, Failures: 2.
```

This is after a fresh `vagrant up --provision`, `composer update`, and making sure my 2.x branch is up to date. On the other hand, Travis looks fine. Have you seen this before? Should I just ignore these and power through?

**Comment by acobster** [comment]
eep! Didn't mean to close this :P

**Comment by acobster** [comment]
Quick update on the unit test failures - apparently it's [failing in Travis](https://travis-ci.org/sitecrafting/timber/jobs/450079911) for the sitecrafting fork as well. Am I missing something?

**Comment by gchtr** [comment]
Thanks for taking this up, @acobster! I think it’s fine if this is a breaking change 👍. That’s why we have the 2.0 label, I guess.

I just want to add another solution that would solve this problem differently: **Object factories**. I mentioned this issue in https://github.com/timber/timber/pull/1219#issuecomment-422441563, and I still think it would actually be the best option.

In terms of testing: I sometimes run into problems like the one you have when I only test one method or only one class. Do you also see the error when you run `phpunit` instead of `phpunit tests/test-timber-post.php`?

**Comment by acobster** [comment]
Yeah, I'm leaning more and more towards just overhauling the semantics and using factories for the creation of everything. I'm going to put this issue on hold and see what I can do to support that effort. If we get that right, handling bad IDs passed to `Timber::get_post()` and friends will naturally fall out of that.

**Comment by gchtr** [comment]
With #2316 merged in, this issue will be resolved with Timber v2. `Timber::get_post()` will return `false` if no post can be found. This means you can use the result of that function directly to check for a valid post:

```php
$post = Timber::get_post();

if ( $post ) {

}
```
